### PR TITLE
Assignment 2

### PR DESCRIPTION
**Pull request by darlingoscanoa**
< UofT-DSI | deep learning - Assignment 2-->

## What changes are you trying to make? (e.g. Adding or removing code, refactoring existing code, adding reports)
We completed from the lab3 t o6
We completed the assignment 2 were we make zero-short transformers to make image classification for the same Fashion MNIST dataset we have used before in the assignment 1 (at that time we made our CNN model) but now we wanted to demostrate the power of transformers methodology. And now using Pytorch instead of Keras.

## What did you learn from the changes you have made?
So many new things:
-How to connect a Hugging Face model with Google Colab (I did almost all the job in colab for the GPU power) 
-Investigate how the models can be public published on HF and request access for them
-Pytorch in the same way as Keras provide some famous dataset in the libraries, and in our case this assignment-2 dont need to train any model so we just download the testing set to evaluate another pretrain model (in this case the openai accesed from HF)
-Zero-shot itself can provide a better performance vs our base line (linear regression) from assignment-1, as openai for sure train its model with a larger dataset.
-How the process works by getting text and image embedding, compute the cosine similarity and also the confusion matrix for these 10 only classes test was great seeing in a visualization the wrong classifications comes from quite similar semantic class names/description 
-Applying some zero-shot prompt engineering increase our performance but not so much.
-Viewing the UMAP was awesome too
-The mini-experiment also provide us a better understanding in my case with top-k how including another closest prediction can increase the performance choose in this case (accuracy) but applying those are related to the use case where some missclassification can be tolerable.

## Was there another approach you were thinking about making? If so, what approach(es) were you thinking of?
Doing the other mini-experiment could increase our chance to improve the metric score.

## Were there any challenges? If so, what issue(s) did you face? How did you overcome it?
Yes, first I tried on my laptop but took more than 20minutes some times, so I moved to Google Colab for the free GPU. I completed the half but after some test the api key stop working and need to solved by clearing the cache from colab with some additional commands once I found the root cause error to continue.

## How were these changes tested?
I checked all steps requested visually and checking the congruency of the results

## A reference to a related issue in your repository (if applicable)
-

## Checklist
- [x ] I can confirm that my changes are working as intended
- [ x] Created a branch with the correct naming convention.
- [ x] Ensured that the repository is public.
- [ x] Reviewed the PR description guidelines and adhered to them.
- [x ] Verify that the link is accessible in a private browser window.

### CODE REVIEW

**Review by DwardEE — APPROVED** [review]
Great job on the assignment! Impressive results with the top-k implmentation at the end. There are no issues with any of the code or answers.
